src/plugins/vis_builder/public/plugin.ts

```diff
@@ -85,28 +135,48 @@ export class VisBuilderPlugin
         // TODO: Add the redirect
         await pluginsStart.data.indexPatterns.ensureDefaultIndexPattern();
 
-        // Register Default Visualizations
+        appMounted();
 
+        // dispatch synthetic hash change event to update hash history objects
+        // this is necessary because hash updates triggered by using popState won't trigger this event naturally.
+        const unlistenParentHistory = params.history.listen(() => {
+          window.dispatchEvent(new HashChangeEvent('hashchange'));
+        });
+
+        // Register Default Visualizations
         const services: VisBuilderServices = {
           ...coreStart,
+          scopedHistory: this.currentHistory,
+          history: params.history,
+          osdUrlStateStorage: createOsdUrlStateStorage({
+            history: params.history,
+            useHash: coreStart.uiSettings.get('state:storeInSessionStorage'),
+            ...withNotifyOnErrors(coreStart.notifications.toasts),
+          }),
           toastNotifications: coreStart.notifications.toasts,
           data,
           savedObjectsPublic: savedObjects,
           navigation,
           expressions,
-          history: params.history,
           setHeaderActionMenu: params.setHeaderActionMenu,
           types: typeService.start(),
           savedVisBuilderLoader: selfStart.savedVisBuilderLoader,
           embeddable: pluginsStart.embeddable,
-          scopedHistory: params.history,
+          setActiveUrl,
+          restorePreviousUrl,
+          dashboard: pluginsStart.dashboard,
```

Comment:
Do we need these services? I dont see them being used anywhere

Reply:
Some of these services are used for restoring history, and make the back button on the browser work. Since they are indeed not related to this issue, will remove them in the services.